cmd/thanos/bucket.go

```diff
@@ -291,6 +299,123 @@ func registerBucketInspect(m map[string]setupFunc, root *kingpin.CmdClause, name
 	}
 }
 
+// RegisterBucketWeb exposes a web interface for the state of Thanos remote store just like `pprof web`
+//
+// TODOS:
+//
+// 1. Some download code is duplicated from `ls -o json`; refactor as needed
+// 2. Some web code is duplicated from query UI, refactor as needed
+// 3. Do we need any of the route prefix?
+func registerBucketWeb(m map[string]setupFunc, root *kingpin.CmdClause, name string, objStoreConfig *pathOrContent) {
+	cmd := root.Command("web", "Web interface for remote storage bucket")
+	bind := cmd.Flag("listen", "HTTP host:port to listen on").Default("0.0.0.0:8080").String()
+	interval := cmd.Flag("refresh", "Refresh interval in minutes").Default("30").Int()
+
+	m[name+" web"] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ bool) error {
+		ctx, cancel := context.WithCancel(context.Background())
+
+		router := route.New()
+
+		bucketUI := ui.NewBucketUI(logger)
+		bucketUI.Register(router)
+
+		g.Add(func() error {
+			return refresh(ctx, logger, bucketUI, int(*interval), name, reg, objStoreConfig)
+		}, func(error) {
+			cancel()
+		})
+
+		l, err := net.Listen("tcp", *bind)
+		if err != nil {
+			return errors.Wrapf(err, "listen HTTP on address %s", *bind)
+		}
+
+		g.Add(func() error {
+			level.Info(logger).Log("msg", "Listening for query and metrics", "address", *bind)
+			return errors.Wrap(http.Serve(l, router), "serve web")
+		}, func(error) {
+			runutil.CloseWithLogOnErr(logger, l, "http listener")
+		})
+
+		return nil
+	}
+}
+
+// Refresh metadata from remote storage periodically and update UI
+func refresh(ctx context.Context, logger log.Logger, bucketUI *ui.Bucket, interval int, name string,
+	reg *prometheus.Registry, objStoreConfig *pathOrContent) error {
+
+	confContentYaml, err := objStoreConfig.Content()
+	if err != nil {
+		return err
+	}
+
+	// A client is created for the entire duration of the program to avoid duplicate metrics registration.
+	bkt, err := client.NewBucket(logger, confContentYaml, reg, name)
+	if err != nil {
+		return err
+	}
+
+	defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+
+	delay := time.Duration(interval) * time.Minute
+
+	return runutil.Repeat(delay, ctx.Done(), func() error {
+		iterCtx, iterCancel := context.WithTimeout(ctx, 5*time.Minute)
```

Comment:
I think we might need to configure timeout at some point as well? Also it's tricky as timeout will mean that we will wait until ... next interval which can be in X minutes? We could simplify flow here a bit I think. I would retry until success OR timeout with hard failure on server. Continuing and waiting for next interval tick is like retrying with backoff to me (:

Reply:
not addressed (:

Reply:
@bwplotka 83dda592358cded322c450a2dda733c435615200 simplifies the loop again and a7012e04be717d5ea91cd20c56b46c9ba89d87d0 made the arguments configurable.